File: mobile_microapps_architecture_flutter/lib/app_shell/app_session_controller.dart
import 'package:flutter/foundation.dart';
import 'package:module_contracts/module_contracts.dart';

final class AppSessionController implements SessionContract {
  AppSessionController()
      : _currentUser = ValueNotifier<SessionUser?>(
          const SessionUser(
            id: 'customer-1001',
            name: 'Gian Sandoval',
            role: 'Enterprise customer',
            segment: 'Premium',
          ),
        );

  final ValueNotifier<SessionUser?> _currentUser;

  @override
  ValueListenable<SessionUser?> get currentUser => _currentUser;

  @override
  SessionUser? get user => _currentUser.value;

  @override
  bool get isAuthenticated => _currentUser.value != null;

  @override
  Future<void> signInAsDemoUser() async {
    _currentUser.value = const SessionUser(
      id: 'customer-1001',
      name: 'Gian Sandoval',
      role: 'Enterprise customer',
      segment: 'Premium',
    );
  }

  @override
  Future<void> signOut() async {
    _currentUser.value = null;
  }
}

File: MobileMicroappsArchitecture/src/app_shell/AppSessionController.ts
import {
  SessionContract,
  SessionListener,
  SessionUser,
} from '../packages/module_contracts';

const demoUser: SessionUser = {
  id: 'customer-1001',
  name: 'Gian Sandoval',
  role: 'Enterprise customer',
  segment: 'Premium',
};

export class AppSessionController implements SessionContract {
  private user: SessionUser | null = demoUser;
  private listeners = new Set<SessionListener>();

  currentUser() {
    return this.user;
  }

  isAuthenticated() {
    return this.user !== null;
  }

  subscribe(listener: SessionListener) {
    this.listeners.add(listener);
    listener(this.user);

    return () => {
      this.listeners.delete(listener);
    };
  }

  async signInAsDemoUser() {
    this.user = demoUser;
    this.notify();
  }

  async signOut() {
    this.user = null;
    this.notify();
  }

  private notify() {
    for (const listener of this.listeners) {
      listener(this.user);
    }
  }
}
